python/veles/schema/fields.py

```diff
@@ -15,214 +15,281 @@
-        if value is None:
-            return
-        if not isinstance(value, bool):
-            raise ValueError('Attribute {} has to be bool type.'.format(
-                self.name))
-        return value
+    value_type = bool
 
 
 class Float(Field):
-    def validate(self, value):
-        super(Float, self).validate(value)
-        if value is None:
-            return
-        if not isinstance(value, float):
-            raise ValueError('Attribute {} has to be float type.'.format(
-                self.name))
-        return value
+    value_type = float
 
 
 class String(Field):
-    def validate(self, value):
-        super(String, self).validate(value)
-        if value is None:
-            return
-        if not isinstance(value, six.text_type):
-            raise ValueError('Attribute {} has to be str type.'.format(
-                self.name))
-        return value
+    value_type = six.text_type
 
 
 class Binary(Field):
-    def validate(self, value):
-        super(Binary, self).validate(value)
-        if value is None:
-            return
-        if not isinstance(value, bytes):
-            raise ValueError('Attribute {} has to be bytes type.'.format(
-                self.name))
-        return value
+    value_type = bytes
 
 
-class Array(Field):
-    def __init__(self, optional=False, elements_types=None, local_type=list):
-        super(Array, self).__init__(optional)
-        self._allowed_local = (list, tuple, set, frozenset)
-        if elements_types:
-            self.elements_types = elements_types
-        else:
-            self.elements_types = [Any()]
-        if local_type not in self._allowed_local:
-            raise ValueError('Illegal local_type value')
-        self.local_type = local_type
+class NodeID(Field):
+    value_type = nodeid.NodeID
 
-    def validate(self, value):
-        super(Array, self).validate(value)
-        if value is None:
-            return []
-        if not isinstance(value, self._allowed_local):
-            raise ValueError(
-                'Attribute {} has to be one of {} type.'.format(
-                    self.name, self._allowed_local))
-        prep_value = []
+
+class BinData(Field):
+    value_type = bindata.BinData
+
+
+class List(Field):
+    def __init__(self, element, optional=False, default=[]):
+        super(List, self).__init__(optional, default)
+        if not isinstance(element, Field):
+            raise TypeError("List element must be a Field.")
+        self.element = element
+
+    def __set_name__(self, owner, name):
+        super(List, self).__set_name__(owner, name)
+        self.element.__set_name__(owner, name + '.element')
+
+    def _validate(self, value):
+        if not isinstance(value, list):
+            raise SchemaError(
+                'Attribute {} has to be a list'.format(self.name))
+        for val in value:
+            self.element.validate(val)
+
+    def _load(self, value):
+        if not isinstance(value, list):
```

Comment:
Right - I forgot that it would call `_validate` from `List`. I wanted to use it to also check `optional`, but overall I think we can get rid of `optional` in `List`, `Set` and `Map` altogether since they aren't used anywhere for those types.
Also we can still use base `_validate` here.